transformations/transformations/data_offload.py

```diff
@@ -176,3 +181,335 @@ def remove_openmp_pragmas(self, routine, targets):
+
+        routine.body = Transformer(pragma_map).visit(routine.body)
+
+        # build set of symbols to be offloaded
+        _var_set = reduce(operator.or_, [s.trafo_data[self._key]['var_set']
+                          for s in successors], set())
+        #build map of module imports corresponding to offloaded symbols
+        _modules = {}
+        _modules.update({k: v
+                         for s in successors if isinstance(s, SubroutineItem)
+                         for k, v in s.trafo_data[self._key]['modules'].items()})
+
+        # build new imports to add offloaded global vars to driver symbol table
+        new_import_map = {}
+        for s in _var_set:
+            if s in routine.symbol_map:
+                continue
+
+            if new_import_map.get(_modules[s], None):
+                new_import_map[_modules[s]] += as_tuple(s)
+            else:
+                new_import_map.update({_modules[s]: as_tuple(s)})
+
+        new_imports = ()
+        for k, v in new_import_map.items():
+            new_imports += as_tuple(Import(k, symbols=tuple(Variable(name=s, scope=routine) for s in v)))
+
+        # add new imports to driver subroutine sepcification
+        import_pos = 0
+        if (old_imports := FindNodes(Import).visit(routine.spec)):
+            import_pos = routine.spec.body.index(old_imports[-1]) + 1
+        if new_imports:
+            routine.spec.insert(import_pos, Comment(text=
+               '![Loki::GlobalVarOffload].....Adding global variables to driver symbol table for offload instructions'))
+            import_pos += 1
+            routine.spec.insert(import_pos, new_imports)
+
+    def process_kernel(self, routine, successors, item, targets):
+        """
+        Collect the set of module variables to be offloaded.
+        """
+
+        # filter out procedure imports
+        calls = [c.name for c in FindNodes(CallStatement).visit(routine.body)]
+        calls += FindInlineCalls().visit(routine.body)
+
+        imports = [i for i in FindNodes(Import).visit(routine.spec) if i.symbols not in calls]
+
+        # build map of modules corresponding to imports
+        import_mod = {s.name.lower(): i.module for i in imports for s in i.symbols}
+
+        #build set of offloaded symbols
+        item.trafo_data[self._key]['var_set'] = reduce(operator.or_,
+                                                       [s.trafo_data[self._key]['var_set'] for s in successors], set())
+
+        #build map of module imports corresponding to offloaded symbols
+        item.trafo_data[self._key]['modules'].update({k: v
+                                                     for s in successors if isinstance(s, SubroutineItem)
+                                                     for k, v in s.trafo_data[self._key]['modules'].items()})
+
+        # separate out derived and basic types
+        basic_types = [s.name.lower() for i in imports for s in i.symbols if s in targets
```

Comment:
Since `imports` likely isn't quite what you want it to be, here another proposal:
```python
imported_vars = [var for var in routine.imported_symbols if var in item.trafo_data[self._key]['var_set']]
basic_types = [var.name.lower() for var in imported_vars if isinstance(var.type.dtype, BasicType)]
deriv_types = [var.name.lower() for var in imported_vars if isinstance(var.type.dtype, DerivedType)]
```